File: frontend/src/lib/ipc/bisect.ts
import { commands, type BisectMark, type RepoId } from "./bindings";
import { unwrap } from "./index";

export type { BisectMark, BisectState } from "./bindings";

/** `git bisect` (doc/04-ipc-contract.md): each step checks out the next commit to test. */
export async function bisectStart(repo: RepoId, bad: string, good: string | null) {
  return unwrap(await commands.bisectStart(repo, bad, good));
}

/** `rev` null marks HEAD, the commit git checked out. */
export async function bisectMark(repo: RepoId, mark: BisectMark, rev: string | null) {
  return unwrap(await commands.bisectMark(repo, mark, rev));
}

export async function bisectReset(repo: RepoId) {
  return unwrap(await commands.bisectReset(repo));
}

File: src-tauri/src/commands/bisect.rs
//! `git bisect`: start, mark a commit, reset (M11). Each step checks out a commit, so it
//! waits in the repository's lane as a checkout does.

use super::mutating_titled;
use app_state::{OperationKind, RepoId};
use git_engine::{BisectMark, GitError};

#[tauri::command]
#[specta::specta]
pub async fn bisect_start(
    state: tauri::State<'_, crate::AppContext>,
    repo: RepoId,
    bad: String,
    good: Option<String>,
) -> Result<(), GitError> {
    let app_state = state.state.clone();
    mutating_titled(
        &state.state,
        repo,
        OperationKind::Checkout,
        "Starting bisect",
        "bisect_start",
        move || app_state.bisect_start(repo, &bad, good.as_deref()),
    )
    .await
}

/// `rev` is HEAD when absent.
#[tauri::command]
#[specta::specta]
pub async fn bisect_mark(
    state: tauri::State<'_, crate::AppContext>,
    repo: RepoId,
    mark: BisectMark,
    rev: Option<String>,
) -> Result<(), GitError> {
    let app_state = state.state.clone();
    let title = match mark {
        BisectMark::Good => "Marking good",
        BisectMark::Bad => "Marking bad",
        BisectMark::Skip => "Skipping",
    };
    mutating_titled(
        &state.state,
        repo,
        OperationKind::Checkout,
        title,
        "bisect_mark",
        move || app_state.bisect_mark(repo, mark, rev.as_deref()),
    )
    .await
}

#[tauri::command]
#[specta::specta]
pub async fn bisect_reset(
    state: tauri::State<'_, crate::AppContext>,
    repo: RepoId,
) -> Result<(), GitError> {
    let app_state = state.state.clone();
    mutating_titled(
        &state.state,
        repo,
        OperationKind::Checkout,
        "Resetting bisect",
        "bisect_reset",
        move || app_state.bisect_reset(repo),
    )
    .await
}
